src/qt/walletmodel.cpp

```diff
@@ -668,3 +686,210 @@ bool WalletModel::saveReceiveRequest(const std::string &sAddress, const int64_t
+    const std::string toaddress = it->second.toaddress;
+
+    updatedNameNewReturn.ok = true;
+    updatedNameNewReturn.hex = txid;
+    updatedNameNewReturn.rand = rand;
+    updatedNameNewReturn.data = strData;
+    if(!toaddress.empty ())
+        updatedNameNewReturn.toaddress = toaddress;
+
+    it->second = updatedNameNewReturn;
+
+    UniValue uniNameUpdateData(UniValue::VOBJ);
+    uniNameUpdateData.pushKV ("txid", txid);
+    uniNameUpdateData.pushKV ("rand", rand);
+    uniNameUpdateData.pushKV ("data", strData);
+    if(!toaddress.empty ())
+        uniNameUpdateData.pushKV ("toaddress", toaddress);
+
+    std::string jsonData = uniNameUpdateData.write();
+    LogPrintf ("Writing name_firstupdate %s => %s\n", strName.c_str(), jsonData.c_str());
+    wallet->WriteNameFirstUpdate(strName, jsonData);
+    return tr("");
+}
+
+void WalletModel::sendPendingNameFirstUpdates()
+{
+    for (std::map<std::string, NameNewReturn>::iterator i = pendingNameFirstUpdate.begin();
+         i != pendingNameFirstUpdate.end(); )
+    {
+        UniValue params1(UniValue::VOBJ), params2(UniValue::VOBJ);
+        UniValue res1, res2, val;
+
+        std::string strName = i->first;
+        std::string txid = i->second.hex;
+        std::string rand = i->second.rand;
+        std::string data = i->second.data;
+        std::string toaddress = i->second.toaddress;
+
+        params1.push_back (Pair("txid", txid));
+
+        // if we're here, the names doesn't exist
+        // should we remove it from the DB?
+        try {
+            res1 = gettransaction( params1, false);
+        }
+        catch (const UniValue& e) {
+            LogPrintf ("gettransaction error for name %s: %s\n", strName.c_str(), e.getValStr().c_str());
+            ++i;
+            continue;
+        }
+
+        val = find_value (res1, "confirmations");
+        if (!val.isNum ())
+        {
+            LogPrintf ("No confirmations for name %s\n", strName.c_str());
+            ++i;
+            continue;
+        }
+
+        const int confirms = val.get_int ();
+        LogPrintf ("Pending Name FirstUpdate Confirms: %d\n", confirms);
+
+        if ( confirms < 12)
+        {
+            ++i;
+            continue;
+        }
+
+        params2 = UniValue::VOBJ;
+        params2.push_back (Pair("name", strName));
+        params2.push_back (Pair("rand", rand));
+        params2.push_back (Pair("tx", txid));
+        params2.push_back (Pair("value", data));
+        if(!toaddress.empty())
+            params2.push_back (Pair("toaddress", toaddress));
+
+        try {
+            res2 = name_firstupdate (params2, false);
+        }
+        catch (const UniValue& e) {
+            LogPrintf ("name_firstupdate error: %s\n", e.getValStr().c_str());
+            ++i;
+            continue;
+        }
+
+        LogPrintf ("sent name_firstupdate %s\n", strName.c_str());
+        pendingNameFirstUpdate.erase(i++);
+        wallet->EraseNameFirstUpdate(strName);
```

Comment:
Just a general comment - it seems that the pending name_firstupdate is only ever deleted from the wallet if it was successfully sent.  Maybe there should be some other mechanism, e. g., if the name was taken by someone else?  Otherwise it may "corrupt" the wallet.  (I know that you just implemented what is in the old code, but maybe we can discuss this and resolve it more thoroughly now or in a future patch.)